### PR TITLE
Remove Interop build step

### PR DESCRIPTION
**Pull request by dougbu**
- remove separate step to restore Interop projects
- make Interop test asset projects more "normal"
    - use usual mechanism to target latest ASP.NET Core bits
    - remove use of `$(BuildInteropProjects)`
- expand FunctionalTestWithAssets.targets to help w/ local testing
    - still delayed and done only when necessary
    - handle non-Helix testing e.g. handle `dotnet test` too
    - build test assets into location `InteropTests` expects on local machine
- add FunctionalTestWithAssets.props, providing relevant defaults
- expand `FunctionalTestWithAssets` use to Identity.FunctionalTests

nits:
- rename `@(FunctionalTestAssetProjectReference)` to `@(TestAssetProjectReference)`
- cleanup leftover reference to ProjectTemplates/testassets/ in Build.props
    - should have been removed in 54ff379
- remove `/p:RunTemplateTests=true`; no need to specify the default
- clean up other projects that use `$(MvcTestingTargets)`
    - none need special handling for publish (don't need pages, views, or wwwroot files)

### CODE REVIEW

File: src/Grpc/test/testassets/InteropWebsite/InteropWebsite.csproj
Comment (dougbu) [inline src/Grpc/test/testassets/InteropWebsite/InteropWebsite.csproj:8]: Note `InteropClient` doesn't depend on Microsoft.AspNetCore.App.Ref

File: src/Grpc/test/InteropTests/InteropTests.csproj
Comment (dougbu) [inline src/Grpc/test/InteropTests/InteropTests.csproj:29]: @rainersigwald why doesn't metadata work correctly when creating an item group using wildcards❔ Things seem to always work when including from another item group or updating one that exists (like the "fixed" case).

Is there any way to set this item group in one go❔

**Comment by JamesNK** [comment]
As long as the gRPC interop tests still run using the latest asp.net core source code then its good with me.

**Comment by dougbu** [comment]
@JamesNK we have lots of experience w/ the approach and I confirmed nothing got in the way of the gRPC interop tests running.

File: eng/targets/FunctionalTestAsset.targets
Comment (HaoK) [inline eng/targets/FunctionalTestAsset.targets:1]: Are we sure we want to delete these?  I believe the idea for these is to generalize what the IIS functionals are doing, so if we were ever going to move the rest of the hosting functionals etc onto helix, we would need something like these.

**Review by wtgodbe — APPROVED** [review]
Seems reasonable, but I'd also like to confirm that we'll never `FunctionTestAsset.targets` or `FunctionalTestWithAssets.targets` again

**Comment by dougbu** [comment]
> Are we sure we want to delete these? I believe the idea for these is to generalize what the IIS functionals are doing, so if we were ever going to move the rest of the hosting functionals etc onto helix, we would need something like these.

> Seems reasonable, but I'd also like to confirm that we'll never `FunctionTestAsset.targets` or `FunctionalTestWithAssets.targets` again

The point of this PR is the files are redundant with the rest of our build infrastructure and were only used for the gRPC interop tests. What requirements do the IIS functionals have that can't be met in other ways❔

**Comment by HaoK** [comment]
Those targets were added by John when he was trying to generalize what the IIS functionals are doing so other tests could do the same thing more easily in the future, so their purposes is general infrastructure for the future, I don't see much value in deleting those targets files and putting the logic back into individual projects, you can still remove the interop step without deleting the shared targets right?

**Comment by dougbu** [comment]
> their purposes is general infrastructure for the future

Note what I put into InteropTests.csproj is simpler than what was in the files before. Much more straightforward to `publish` where things are needed than to `publish` then `<Copy/>`.

> putting the logic back into individual projects

I wasn't proposing putting the logic back into any individual project except InteropTests.csproj. I could however duplicate the small bits needed in FunctionalTests.props (or, better, a new FunctionalTests.targets since it contains `<Target/>`s and not much else). Other options include leaving this PR as-is and simplifying the IIS functional tests later.

Offline, I proposed restoring the central files, but I don't really like that option (even with simplifications) because they're useless for almost all of our tests.

**Comment by HaoK** [comment]
Pushing the improvements into the central bits sounds like a win to me, my only pushback was against deleting the centralized functional tests targets/props since the existing IIS functionals are not generalized at all, hopefully we can move them (and convert the hosting functionals) onto these in some future MQ milestone

**Comment by dougbu** [comment]
@HaoK I'm not understanding why centralizing the simple `PublishTestAssets` target anywhere is a "win". We centralize things when they are broadly useful, but don't when only small areas need the infrastructure e.g. the specialized infrastructure for project template tests is _not_ in our eng folder. I don't see any harm in having the same target in InteropTests.csproj, FunctionalTests.targets, and somewhere else for the hosting tests.

Of the options I listed, which can you live with❔

**Comment by HaoK** [comment]
One of the goals of the work in the InteropTests was to generalize the helix publishing stuff, which is why there was a creation of these centralized functional test props/assets.  I wasn't too involved in the specifics here, but if the result of this PR is now that we just have the interop tests doing their own thing, as well as the IIS functionals doing their own thing, we are no closer to having all of our functional tests sharing the same helix/publishing infrastructure.  If you feel like your new pattern is an improvement on what was there before, then I think it'd be sufficient to document it in helix.md I guess, I mostly just want to make sure we don't lose whatever work John did to generalize things into shared files.  The problem today is that its clearly not trivial to publish/run our functionals on helix because we haven't gotten them all running and we have at least two different ways to do it, and every time someone does it, they roll something slightly new.  I just want to make sure we aren't going backwards, and instead are moving towards a common pattern

**Comment by HaoK** [comment]
Maybe before merging we should also make sure things are being properly tested by temporarily introducing an actual break in this PR somewhere in the interop codebase that the interop tests guard against to ensure the full end to end test flow still works

File: src/Servers/Kestrel/Core/src/KestrelServerLimits.cs
Comment (BrennanConroy) [inline src/Servers/Kestrel/Core/src/KestrelServerLimits.cs:46]: ???

Comment (dougbu) [reply src/Servers/Kestrel/Core/src/KestrelServerLimits.cs:46]: This isn't for realz. I'm just adding a change that should break the gRPC interop tests, demonstrating they run against the just-built Microsoft.AspNetCore.App shared Fx bits.

**Comment by dougbu** [comment]
With my runtime break included, interop tests are all failing. I busted some Kestrel tests too (of course). The `InteropTests` job on macOS looks likely to time out before reporting the same problems. Is that enough confirmation to go back to addressing your other comments @HaoK❔

----

Should have mentioned that I confirmed the tests were running and _un_broken prior to introducing the `!do not merge!` commit. We have other confirmations in place, especially in template tests, confirming the latest ASP.NET shared Fx is available and used on Helix agents.

**Comment by HaoK** [comment]
Yep looks like the behavior isn't broken, so its just a matter of the other questions

**Comment by dougbu** [comment]
Where would the proposed annotations be visible❔

In my recent case (discussed in FR), these were rolling builds; the `Build Analysis` tab wouldn't help in that case.

**Comment by dougbu** [comment]
After spending a while on this, I'd say there's pretty much no commonality between the gRPC tests and IIS tests. In the gRPC case, delaying test asset builds until after Microsoft.AspNetCore.App.Ref and &hellip;Runtime are built and placing build and publish output directly under the test project's bin/ folder makes sense. Neither is sensible for IIS because multiple test projects depend on the same test asset and none of the test asset projects need the latest shared Fx.

The current FunctionalTestAsset.targets and FunctionalTestWithAssets.targets files in main also don't make much sense for either of these cases. The approach is inefficient (resulting in useless copies) for gRPC and not extensible enough for IIS.

I'm tempted to build IIS test projects like any other project reference but use a subset of the new approach in this PR to get the bits into their final locations. In particular, build wouldn't change (other than not automatically triggering publication of the test assets) but publish and test-related targets would `Publish` w/ `NoBuild = true` under the test project's bin/ folder. This would however result in duplicate `Publish` operations and `Publish` does a lot of analysis before copying files around.

The other option is to give up on centralizing things and let gRPC and IIS be unique. Each can be improved e.g. FunctionalTest.props isn't perfect and `AfterTargets` is unreliable, but they won't improve together.

All that said, Microsoft.AspNetCore.Identity.FunctionalTests.csproj might be able to share a few bits w/ the gRPC test project. Most of the copies done there would actually be better handled as `<None ... CopyToPublishDirectory="PreserveNewest" Visible="false" />` or `<Content ... />` items.

Thoughts @wtgodbe @HaoK❔

**Comment by dougbu** [comment]
Separately, @rainersigwald @Rick-Anderson I see https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2022 documents `%(CopyToOutputDirectory)` metadata but doesn't mention `%(CopyToPublishDirectory)`. Where would the issue mentioned in MicrosoftDocs/visualstudio-docs#53 be found❔

**Comment by HaoK** [comment]
I think if IIS/gRPC are sufficiently unique, lets leave them be

But if you can standardize other functional tests (Identity and AuthSamples both do the same thing), that seems worth doing.

**Comment by dougbu** [comment]
:ok: and thanks for the quick response.

I'll expand the bits I have for gRPC to cover another case or two but leave the IIS projects alone for now. I might prioritize making the IIS projects more "normal" e.g. place bin/ folders in their normal locations over improving FunctionalTest.props (or even renaming that "targets" file 😺).

**Comment by dougbu** [comment]
> I might prioritize

To be clear, I'm not taking ownership of IIS test improvements.

**Comment by Rick-Anderson** [comment]
> Separately, @rainersigwald @Rick-Anderson I see https://docs.microsoft.com/en-us/visualstudio/msbuild/common-msbuild-project-items?view=vs-2022 documents `%(CopyToOutputDirectory)` metadata but doesn't mention `%(CopyToPublishDirectory)`. Where would the issue mentioned in [MicrosoftDocs/visualstudio-docs#53](https://github.com/MicrosoftDocs/visualstudio-docs/issues/53) be found❔

@Mikejo5000 @ghogen

File: src/Mvc/test/Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj
Comment (pranavkm) [inline src/Mvc/test/Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj:20]: The older thing kinda looked nicer at the expense of being more verbose. Does specifying things this way help in some way?

Comment (dougbu) [reply src/Mvc/test/Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj:20]: I think we have different ideas of niceness😃 I only did this to reduce verbosity and get w/ the program of SDK projects shortening project files. Can revert if I'm offending your 👀

Comment (dougbu) [reply src/Mvc/test/Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj:20]: @pranavkm please let me know whether this was a nit or not.

**Comment by dougbu** [comment]
@HaoK @wtgodbe please re-review. Things are different from when you last commented or approved.

File: src/Mvc/test/Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj
Comment (pranavkm) [reply src/Mvc/test/Mvc.FunctionalTests/Microsoft.AspNetCore.Mvc.FunctionalTests.csproj:20]: I'll resolve this. Just curious more than anything.

**Review by wtgodbe — APPROVED** [review]

**Comment by dougbu** [comment]
Thanks @wtgodbe. @HaoK any remaining issues❔

**Review by HaoK — APPROVED** [review]
Nice to see some of the test publish jank get cleaned up!

**Pull request merged**
